### PR TITLE
Codechange: Make window widget lists build-time.

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

Window NWidgetPart lists are const, but are actually built on start up because NWidgetPart functions are not constexpr.

Making NWidgetParts constexpr means that runtime variables cannot be used inadvertently.

## Description

WIdgetDimensions::unscaled needs to be moved to the header file to also be constexpr, C++17 should support this without making multiple copies per file that includes it.

Add constexpr NWidgetPart constructors, and make NWidgetPart factory functions constexpr, and make NWIdgetPart lists constexpr too.

## Limitations

Whew

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/window_gui.h [hunk 1]
```diff
@@ -65,6 +66,151 @@ struct WidgetDimensions {
...
 	static WidgetDimensions scaled;         ///< Widget dimensions scaled for current zoom level.
+
+private:
+	/**
+	 * Distances used in drawing widgets.
+	 * These constants should not be used elsewhere, use scaled/unscaled WidgetDimensions instead.
+	 */
+	enum WidgetDrawDistances {
+		/* WWT_IMGBTN(_2) */
+		WD_IMGBTN_LEFT    = 1,      ///< Left offset of the image in the button.
```
Comment (TrueBrain) [inline src/window_gui.h:77]: Now the `WD_` prefix is a bit unneeded. Maybe not for this PR :)

Comment (PeterN) [reply src/window_gui.h:77]: To be honest most of these can be removed anyway and replaced with literal numbers in the unscaled definitions. Like I've been doing for new entries.

File: src/core/geometry_type.hpp [hunk 2]
```diff
@@ -57,13 +57,13 @@ struct RectPadding {
 	 * Get total horizontal padding of RectPadding.
 	 * @return total horizontal padding.
 	 */
-	inline uint Horizontal() const { return this->left + this->right; }
+	inline constexpr uint Horizontal() const { return this->left + this->right; }
```
Comment (TrueBrain) [inline src/core/geometry_type.hpp]: `[dcl.constexpr], §7.1.5/2 in the C++11 standard): "constexpr functions and constexpr constructors are implicitly inline (7.1.2)`.

Could drop the `inline`. Nitpicking ftw :)

File: src/window_gui.h (hunk 1 above)
Comment (PeterN) [reply src/window_gui.h:77]: Hmm... looks kinda random... https://gist.github.com/PeterN/d083990f8445f71b4828974b48f80fd0

Comment (rubidium42) [reply src/window_gui.h:77]: I do like https://gist.github.com/PeterN/d083990f8445f71b4828974b48f80fd0?permalink_comment_id=4834397#gistcomment-4834397. It's quite self explanatory.
Except that left/right/top/bottom is not directly clear. Maybe those could be added for completeness too? But that might clutter it a lot too, so... don't really know.

File: src/core/geometry_type.hpp (hunk 2 above)
Comment (PeterN) [reply src/core/geometry_type.hpp]: Done, there was quite a lot added, so worth it :)

**Review by TrueBrain — APPROVED** [review]
The open comment can be picked up in another PR. Your commit just moves code. So I see no reason for that to be a blocker for merging this.

**Pull request merged**